kubeadm: use DefValue for the --kubeconfig flag

neolit123 opened the pull request:
**What this PR does / why we need it**:

- Rename FindExistingKubeConfig to FindKubeConfigFile

- Cobra supports a DefValue option which can be used
to differentiate between the cases where the user set a flag
and when a flag was unset, while still adding a default value.
Use this in options/generic.go for the kubeconfig flag.

- Remove the FindKubeConfigFile() logic from `reset` and `upgrade`
as these are node level kubeadm commands.

- Default kubeconfig values to "" everywhere where FindKubeConfigFile
is used. This allows to search for existing kubeconfig locations.

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes kubernetes/kubeadm#1304

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
kubeadm: fix a bug where standard kubeconfig paths were searched even if the user provided /etc/kubernetes/admin.conf explicitly for commands that accept --kubeconfig, like kubeadm token.
kubeadm: use the default kubeconfig (/etc/kubernetes/admin.conf) for "kubeadm reset" and "kubeadm upgrade" commands.
```

/assign @fabriziopandini 
/cc @yagonobre 
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews 
/priority important-longterm
/hold

yagonobre commented on cmd/kubeadm/app/cmd/util/cmdutil.go:
```diff
@@ -68,8 +68,9 @@ func FindExistingKubeConfig(file string) string {
+	// as the primary location to check.
 	rules := clientcmd.NewDefaultClientConfigLoadingRules()
-	rules.Precedence = append(rules.Precedence, kubeadmconstants.GetAdminKubeConfigPath())
+	rules.Precedence = append([]string{kubeadmconstants.GetAdminKubeConfigPath()}, rules.Precedence...)
```
IMO the `KUBECONFIG` env var should take precedence over the default path, so we need to modify `NewDefaultClientConfigLoadingRules`

neolit123 replied:
that was my original thinking when i added this code as well.
but i think @fabriziopandini disagreed with that the last time we spoke about it.

we cannot easily modify `NewDefaultClientConfigLoadingRules` as it might break client-go for multiple consumers. we can hack around it though, by finding where the KUBECONFIG env var is in the slice and prepending. not a great hack.

yagonobre replied:
we can check if the first element of the slice is `RecommendedHomeFile` and if yes prepending the default path, if not append to the end of slice.

neolit123 replied:
ah, i guess this will work yes.
will still wait on @fabriziopandini to see what he thinks about "kubeadm default path vs KUBECONFIG env" as the first search location.

also:
/assign @timothysc

mauilion commented:
+1 to `KUBECONFIG` precedence. This is what `kubectl` does so that's likely the expectation

fabriziopandini commented:
As far as I'm concerned the main use case for kubeadm is to take care of the node that exist locally on the machine.
This is e.g. the case of `kubeadm init`,  and if I remember well what was discussed with @luxas when implementing `kiubeadm init phases` we agreed that use *only* the kubeadm path for those commands in order to avoid the user mess up other machine by mistake

`Kubeadm reset` in my opinion should do the same (use only kubeadm default path)

The kubeconfig path should be considered only in few utilities command that the users might run on its own (without a node on the machine) e.g. `kubeadm token` or `kubeadm config default`, but IMO also in those cases kubeadm default path should go first: if there is a node one the machine, most probably this is the one I want to manage....

neolit123 commented:
^ we need to decide on the kubeconfig search precedence.
/assign @timothysc @luxas 

1) have the search order the same as kubectl - search the KUBECONFIG env var first
2) search for the kubeadm default kubeconfig path first - /etc/kubernetes/admin.conf

neolit123 commented:
@fabriziopandini the majority seem to suggest that we should search KUBECONFIG first.
@luxas do you have an opinion?

rosti commented:
In my opinion, we should strive to reduce the confusion and act like kubectl. KUBECONFIG should take precedence over the default path.

fabriziopandini commented:
@neolit123 I start getting the impression that changing this without a master plan in the background that considers all the kubeadm commands is potentially risky

Prior art:  after discussion with @luxas and you when implementing https://github.com/kubernetes/kubernetes/pull/70763, using KUBECONFIG first was dropped because of the potential risks, and as a consequence the current behavior for `kubeadm init` and `kubeadm init phases` does not consider KUBECONFIG at all.

Now we are changing opinion for `kebadm reset`...
Might be better to discuss this at sig-level

neolit123 commented:
> Prior art: after discussion with @luxas and you when implementing #70763, using KUBECONFIG first was dropped because of the potential risks, and as a consequence the current behavior for kubeadm init and kubeadm init phases does not consider KUBECONFIG at all.

i'm mostly indifferent about the search order myself.
but i see others have an opinion here.

> Now we are changing opinion for kebadm reset...

the cross linked issue is about `kubeadm token`.

> Might be better to discuss this at sig-level

we can also ask the #kubeadm channel with a vote.
but let's talk about it on wednesday.

timothysc commented:
Let's clear this one out this week.

neolit123 commented:
/remove-kind bug
/kind cleanup

updated this PR to only remove the search precedence on `reset`.
other locations where precedence is used:
```
app/cmd/config.go:86:	kubeConfigFile = cmdutil.FindExistingKubeConfig(kubeConfigFile)
app/cmd/token.go:124:			kubeConfigFile = cmdutil.FindExistingKubeConfig(kubeConfigFile)
app/cmd/token.go:151:			kubeConfigFile = cmdutil.FindExistingKubeConfig(kubeConfigFile)
app/cmd/token.go:175:			kubeConfigFile = cmdutil.FindExistingKubeConfig(kubeConfigFile)
app/cmd/upgrade/upgrade.go:65:	flags.kubeConfigPath = cmdutil.FindExistingKubeConfig(flags.kubeConfigPath)
app/cmd/alpha/selfhosting.go:122:			kubeConfigFile = cmdutil.FindExistingKubeConfig(kubeConfigFile)
app/cmd/alpha/kubelet.go:151:			kubeConfigFile = cmdutil.FindExistingKubeConfig(kubeConfigFile)
app/cmd/alpha/certs.go:171:		kubeConfigPath := cmdutil.FindExistingKubeConfig(cfg.kubeconfigPath)
```

search precedence might make sense for these.

current precedence for the above commands:
- explicit path
- KUBECONFIG env variable paths
- user home directory `~/.kube/...`
- default kubeadm path `etc/kubernetes/admin.conf`

closed this issue as not a bug:
https://github.com/kubernetes/kubeadm/issues/1304

ping @fabriziopandini and @timothysc for LGTM.

neolit123 commented:
/retest

MalloZup commented on cmd/kubeadm/app/cmd/util/cmdutil.go:
```diff
@@ -62,15 +62,16 @@ func ValidateExactArgNumber(args []string, supportedArgs []string) error {
+			return value
+		}
 	}
-	// The user did not provide a --kubeconfig flag. Find a config in the standard
```
Nit : I personally would keep this comment

neolit123 replied:
well this is sort of covered in the comment above the function now.

MalloZup reviewed (approved):
+1

neolit123 commented on cmd/kubeadm/app/cmd/util/cmdutil.go:
```diff
@@ -62,15 +62,16 @@ func ValidateExactArgNumber(args []string, supportedArgs []string) error {
+// FindExistingKubeConfig returns the location of a kubeconfig file.
+// If the user passed the --kubeconfig flag respect it, otherwise search for existing
+// kubeconfig files in default locations.
+func FindExistingKubeConfig(flags *pflag.FlagSet) string {
```
if we check over a string, there is no way to determine if the user set the flag or if it was defaulted to a value by cobra.

in the linked issue the user was passing: `--kubeconfig=/etc/kubernetes/admin.conf` and it still used the KUBECONFIG env.

flags.Changed() solves this problem.

rosti replied:
We must not default the flag in Cobra. We have to make sure it's defaulted to an empty string.

rosti replied:
This is kind of the case of the `cri-socket` flag.

neolit123 replied:
here is an example of what happens in kubeoconfig.go:
https://github.com/kubernetes/kubernetes/blob/639dae9a9a6602ac194a171744cf84bdcba9f29a/cmd/kubeadm/app/cmd/config.go#L65-L84
- we default a local variable to `/etc/kubernetes/....`
- then use `options..` to add the flag.
- at this point the default becomes `/etc/kubernetes/...`

how should this be handled instead?
the CRI flag passes "" as the default.
https://github.com/kubernetes/kubernetes/blob/b5b627d522ed74549e2583b258a7df0e3ffd33ea/cmd/kubeadm/app/cmd/reset.go#L88

problem is we want to show `/etc/kubernetes...` as the default path when the user types `kubeadm... --help`.

rosti replied:
`pflag` supports special field in its `Flag` struct called `DefValue` ([see here](https://github.com/spf13/pflag/blob/24fa6976df40757dce6aea913e7b81ade90530e1/flag.go#L176)). It's used solely for printing the default value for usage information and not for initialization. You can overwrite that just after flag insertion with something along those lines:

```
func AddKubeConfigFlag(fs *pflag.FlagSet, kubeConfigFile *string, defaultKubeConfigFile string) {
	fs.StringVar(kubeConfigFile, KubeconfigPath, "", "The kubeconfig file to use when talking to the cluster. If the flag is not set, a set of standard locations are searched for an existing KubeConfig file.")
	fs.Lookup(KubeconfigPath).DefValue = defaultKubeConfigFile
}
```

Then the invocation will be something along those lines:
```
kubeConfigPath := ""
// ....
options.AddKubeConfigFlag(cmd.Flags(), &kubeConfigPath, constants.GetAdminKubeConfigPath())
```

In theory this should work, but I did not test it.

P.S. You need the `Lookup` call, because `StringVar` does not return a `*pflag.Flag`.

neolit123 replied:
> `pflag` supports special field in its `Flag` struct called `DefValue`

ok i didn't know this existed.
will give this a test in the following days.

neolit123 committed:
kubeadm: use DefValue for the --kubeconfig flag

- Rename FindExistingKubeConfig to GetKubeConfigPath

- Cobra supports a DefValue option which can be used
to differentiate between the cases where the user set a flag
and when a flag was unset, while still adding a default value.
Use this in options/generic.go for the kubeconfig flag.

- Remove the GetKubeConfigPath() logic from `reset` and `upgrade`
as these are node level kubeadm commands.

- Default kubeconfig values to "" everywhere where GetKubeConfigPath
is used. This allows to search for existing kubeconfig locations.

neolit123 commented on cmd/kubeadm/app/cmd/reset.go:
```diff
@@ -61,10 +61,12 @@ func NewCmdReset(in io.Reader, out io.Writer) *cobra.Command {
 				kubeadmutil.CheckErr(err)
+				klog.V(1).Infof("[reset] loaded client set from kubeconfig file: %s", kubeConfigFile)
+			} else {
+				klog.V(1).Infof("[reset] could not get client set from missing kubeconfig file: %s", kubeConfigFile)
```
added some V(1) output for this..

neolit123 commented:
@fabriziopandini @rosti @bart0sh 
updated the PR to follow the last suggestion by @rosti for Cobra's DefValue.
the OP includes details of what this PR is doing as a summary.

summary of how kubeconfig loading works currently:
- init (phases),  join (phases), reset & upgrade, use the default path only.
the way this is done - always preset the kubeconfig to the default value before adding the flag using `AddKubeConfigFlag`. don't call `FindKubeConfigFile`.
- for other commands if a variable with no value is passed to `AddKubeConfigFlag(` this value will end up in `FindKubeConfigFile` and default locations will be searched. however if the user sets a value for that flag the value will reach `FindKubeConfigFile` and the function will act as a no-op.

i don't think i can improve the logic more here, so let's try to settle on a decision for 1.14.

bart0sh commented on cmd/kubeadm/app/cmd/util/cmdutil.go:
```diff
@@ -62,18 +63,21 @@ func ValidateExactArgNumber(args []string, supportedArgs []string) error {
+	file = rules.GetDefaultFilename()
+	klog.V(1).Infof("Using kubeconfig file: %s", file)
+	return file
 }
```
I still find It confusing that this machinery can return path to the file that doesn't exist. Should we check it here and return an error or at least warn user somehow?

neolit123 replied:
this value will eventually run a function that tries to open this file and an error will be thrown if the file is missing.

bart0sh replied:
The point is that possible error is not handled immediately. We (intentionally?) leave error handling for future processing which may or not may happen. This is generally fragile and error prone approach in my opinion.

neolit123 replied:
please show examples in the source code where error handling is delayed for kubeconfig.

bart0sh replied:
Sure. Almost any place where FindExistingConfig function is called, e.g. https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/alpha/certs.go#L171
The potential error (returned path to not existing config) is not processed in the FindExistingConfig function. It's left unprocessed in a hope that it will be processed in kubeconfigutil.ClientSetFromFile function.

neolit123 replied:
and ClientSetFromFile will fail if the file is missing.

the function calls:
https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/staging/src/k8s.io/client-go/tools/auth/clientauth.go#L88
which already has a check if a file exists.

we could add another check in FindKubeConfigFile but it would be essentially a double check for all our code.

now, if this function is user facing in a way where the user should always receive an existing file path as a "string" that is printed in the terminal, then it would make sense to modify this function. except that's not what we are doing in our codebase today.

bart0sh replied:
up to you. In my opinion it would make sense to catch potential errors as soon as possible. Current code doesn't do it, leaving error processing to LoadFromFile, ClientSetFromFile, getClientset and other functions. That processing may or may not happen. I hope it happens in current Kubeadm code. However, looking at the calls of FindExistingKubeConfig in  app/cmd/upgrade/upgrade.go and a couple of similar places I would't be surprised if it doesn't happen at all.

bart0sh replied:
@fabriziopandini, @rosti What do you guys think about this? I can explain it in more details if needed.

rosti replied:
Depends on the use cases, which TBH, I did not check in their entirety. If we always load that file after we call this function and not do anything else with it prior to that, then we are OK. Otherwise, it will not be very safe.
If we don't need an extra check, let's not have it, but instead, we can rename the func to something, that implies getting a config file path, rather than verifying it exists. In that case words like "find" and "existing" should not be included in the function name.

Considering this, how about `s/FindKibeConfigFile/GetKubeConfigPath`?

neolit123 replied:
the rename SGTM will update.

neolit123 replied:
i just had a look and all calls to (the newly renamed) `GetKubeConfigPath` results in a kubeconfig string that is always passed to `ClientSetFromFile`.

which to me confirms this:
> If we always load that file after we call this function and not do anything else with it prior to that, then we are OK. Otherwise, it will not be very safe.

rosti commented on cmd/kubeadm/app/cmd/options/generic.go:
```diff
@@ -19,11 +19,13 @@ package options
 func AddKubeConfigFlag(fs *pflag.FlagSet, kubeConfigFile *string) {
-	fs.StringVar(kubeConfigFile, KubeconfigPath, *kubeConfigFile, "The kubeconfig file to use when talking to the cluster. If the flag is not set, a set of standard locations are searched for an existing KubeConfig file.")
+	fs.StringVar(kubeConfigFile, KubeconfigPath, *kubeConfigFile, "The kubeconfig file to use when talking to the cluster. If the flag is not set, a set of standard locations can be searched for an existing kubeconfig file.")
+	fs.Lookup(KubeconfigPath).DefValue = constants.GetAdminKubeConfigPath()
```
Though I consider this the right thing to do ATM, it is a bit hacky, so let's put a comment of some sorts here about a possible difference on help message default value and the actual default value.

neolit123 replied:
comment added.

neolit123 commented:
@rosti @bart0sh @timothysc @fabriziopandini et al
code freeze is coming soon. are we blocking on this for 1.14?

bart0sh commented:
/lgtm

neolit123 commented:
this PR is approved and has thumbs up by @bart0sh and @rosti 
final call on reviews before `/hold cancel` this weekend.

neolit123 commented:
/hold cancel
